### PR TITLE
fix: Update config.go and framework_config.go to use transport.Creds for all cases

### PR DESCRIPTION
**Pull request by andyrzhao**
This PR fixes the issue in https://github.com/hashicorp/terraform-provider-google/issues/14411 by ensuring we use transport.Creds for initializing the credentials/tokenSource in all applicable code-path. Using transport.Creds ensures that the underlying OAuth2 HTTPClient used for token exchange is properly configured for mTLS - the current version of the code in terraform bypasses this logic when calling the oauth2 google package APIs directly.

What is not included in this PR:
1. Removing the custom tokenSource override logic - this is because Terraform persists the custom tokenSource for use in other locations such as for initializing big table grpc client as well as populating data source config for GKE clusters. The transport API for creating HTTPClient does not export the underlying tokenSource for re-use. (This support may be added in a future version of the API, which is currently being re-designed.)
2. Simplifying the duplicated ADC logic in terraform. We will revisit this in the future.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:bug
Update config.go and framework_config.go to use transport.Creds for all cases
```

### CODE REVIEW

**Comment by rileykarson** [comment]
@SarahFrench: I've been working w/ @andyrzhao internally, I'll grab this review!

File: mmv1/third_party/terraform/framework_utils/framework_config.go.erb [hunk 1]
```diff
@@ -601,7 +601,7 @@ func GetCredentials(ctx context.Context, data ProviderModel, initialCredentialsO
 			return *creds
 		}
 
-		creds, err := googleoauth.CredentialsFromJSON(ctx, []byte(contents), clientScopes...)
+		creds, err := transport.Creds(ctx, option.WithCredentialsJSON([]byte(contents)), option.WithScopes(clientScopes...))
```
Comment (rileykarson) [inline mmv1/third_party/terraform/framework_utils/framework_config.go.erb:604]: Note: `transport.Creds` ultimately calls out to `CredentialsFromJSONWithParams`, so this is almost certainly compatible: https://github.com/googleapis/google-api-go-client/blob/cf0df64489c5086a57b9156d9648fe8342b22408/internal/creds.go#LL108C1-L108C1

File: mmv1/third_party/terraform/framework_utils/framework_config.go.erb [hunk 2]
```diff
@@ -625,14 +625,12 @@ func GetCredentials(ctx context.Context, data ProviderModel, initialCredentialsO
 
 	tflog.Info(ctx, "Authenticating using DefaultClient...")
 	tflog.Info(ctx, fmt.Sprintf("  -- Scopes: %s", clientScopes))
-	defaultTS, err := googleoauth.DefaultTokenSource(context.Background(), clientScopes...)
+	creds, err := transport.Creds(context.Background(), option.WithScopes(clientScopes...))
```
Comment (rileykarson) [inline mmv1/third_party/terraform/framework_utils/framework_config.go.erb:628]: Note: `DefaultTokenSource` calls `FindDefaultCredentials` and extracts the TokenSource from the credenetials (which we then use to reconstruct a credentials object...), Creds calls `FindDefaultCredentials` and returns the credentials.

**Review by rileykarson — COMMENTED** [review]
Cool- these interfaces ultimately call the same methods we used to, preserving more metadata throughout. I can't think of a reason that this wouldn't be backwards-compatible.

Would it be easy to add a test that failed before but succeeds now, or is configuring making mtls calls difficult?

**Comment by andyrzhao** [comment]
> Cool- these interfaces ultimately call the same methods we used to, preserving more metadata throughout. I can't think of a reason that this wouldn't be backwards-compatible.
> 
> Would it be easy to add a test that failed before but succeeds now, or is configuring making mtls calls difficult?

Thanks for the review Riley! So in terms of test coverage, adding a unit test (such as verifying that the resulting tokenSource has an oauth2 http client configured if mTLS is enabled) does not seem appropriate since we would be testing the internal behavior of transport.Creds which is subject to change. An integration test that can verify the end-to-end behavior would be more suitable - it would involve a test user account configured with CAA policy making a test call to GCP APIs. I'm not familiar enough with Terraform to make a call on how easy it would be to set-up such a test. From CBA team perspective, we do have plans in the back log to add automated E2E test coverage for all OS's and client tools.

I did manually verify the 2 code paths I touched with this PR (by using the application default credentials, as well as by specifying the GOOGLE_CREDENTIALS override), and both now works with CAA.

File: mmv1/third_party/terraform/framework_utils/framework_config.go.erb (hunk 1 above)
[the remark this answers is not in the document]
Comment (rileykarson) [reply mmv1/third_party/terraform/framework_utils/framework_config.go.erb:604]: Ah- this wasn't it

File: mmv1/third_party/terraform/transport/config.go.erb [hunk 3]
```diff
@@ -1183,15 +1183,12 @@ func (c *Config) GetCredentials(clientScopes []string, initialCredentialsOnly bo
...
+	creds, err := transport.Creds(context.Background(), option.WithScopes(clientScopes...))
 	if err != nil {
 		return googleoauth.Credentials{}, fmt.Errorf("Attempted to load application default credentials since neither `credentials` nor `access_token` was set in the provider block.  No credentials loaded. To use your gcloud credentials, run 'gcloud auth application-default login'.  Original error: %w", err)
 	}
 
-	return googleoauth.Credentials{
-		TokenSource: defaultTS,
-	}, err
-}
+	return *creds, nil
```
Comment (rileykarson) [inline mmv1/third_party/terraform/transport/config.go.erb:1191]:
```suggestion
	return *creds, nil
}
```

File: mmv1/third_party/terraform/transport/config.go.erb [hunk 4]
```diff
@@ -1161,7 +1161,7 @@ func (c *Config) GetCredentials(clientScopes []string, initialCredentialsOnly bo
 			return *creds, nil
 		}
 
-		creds, err := googleoauth.CredentialsFromJSON(c.Context, []byte(contents), clientScopes...)
+		creds, err := transport.Creds(c.context, option.WithCredentialsJSON([]byte(contents)), option.WithScopes(clientScopes...))
```
Comment (rileykarson) [inline mmv1/third_party/terraform/transport/config.go.erb:1164]:
```suggestion
		creds, err := transport.Creds(c.Context, option.WithCredentialsJSON([]byte(contents)), option.WithScopes(clientScopes...))
```

**Review by rileykarson — APPROVED** [review]

**Review by rileykarson — APPROVED** [review]

**Pull request merged**